File: CodingInterviews/python/get_next_from_tree.py
# -*- coding:utf-8 -*-
class TreeLinkNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.next = None

class Solution:
    def GetNext(self, pNode):
        if pNode is None:
            return None
        if pNode.right is not None:
            pNode = pNode.right
            while pNode.left is not None:
                pNode = pNode.left
            return pNode
        else:
            while True:
                tempNode = pNode
                pNode = pNode.next
                if pNode is None or pNode.left == tempNode:
                    return pNode

if __name__ == '__main__':
    solution = Solution()
    print solution.StrToInt('-123')

File: CodingInterviews/java/GetNextFromTree.java
package algorithm;

class TreeLinkNode {
	int val;
	TreeLinkNode left = null;
	TreeLinkNode right = null;
	TreeLinkNode next = null;

	TreeLinkNode(int val) {
		this.val = val;
	}
}

public class GetNextFromTree {
	public TreeLinkNode GetNext(TreeLinkNode pNode) {
		if (pNode == null)
			return null;
		if (pNode.right != null) {
			pNode = pNode.right;
			while (pNode.left != null)
				pNode = pNode.left;
			return pNode;
		} else {
			do {
				TreeLinkNode tempNode = pNode;
				pNode = pNode.next;
				if (pNode == null || pNode.left == tempNode)
					return pNode;
			} while (true);
		}
	}
}